[Merged by Bors] - GitHub Action: Check local Markdown links on push

FlyingRatBull opened the pull request:
Adds an GitHub Action to check all local (non http://, https:// ) links in all Markdown files of the repository for liveness.
Fails if a file is not found.

# Goal
This should help maintaining the quality of the documentation.

# Impact
Takes ~24 seconds currently and found 3 dead links (pull requests already created).

# Dependent PRs
* #2064 
* #2065 
* #2066

# Info
See [markdown-link-check](https://github.com/marketplace/actions/markdown-link-check).

# Example output
```
FILE: ./docs/profiling.md

1 links checked.

FILE: ./docs/plugins_guidelines.md

37 links checked.

FILE: ./docs/linters.md
[✖] ../.github/linters/markdown-lint.yml → Status: 400 [Error: ENOENT: no such file or directory, access '/github/workspace/.github/linters/markdown-lint.yml'] {
  errno: -2,
  code: 'ENOENT',
  syscall: 'access',
  path: '/github/workspace/.github/linters/markdown-lint.yml'
}
```

# Improvements
* Can also be used to check external links, but fails because of:
  * Too many requests (429) responses:
```
FILE: ./CHANGELOG.md
[✖] https://github.com/bevyengine/bevy/pull/1762 → Status: 429
```
   * crates.io links respond 404
```
FILE: ./README.md
[✖] https://crates.io/crates/bevy → Status: 404
```

bjorn3 commented on .github/workflows/action.yml:
```diff
@@ -0,0 +1,14 @@
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@master
+    - uses: gaurav-nelson/github-action-markdown-link-check@v1
```
Can you pin a specific commit? That reduces the risk of a malicious update.

FlyingRatBull replied:
Done, pinned to current version ([1.0.12](https://github.com/gaurav-nelson/github-action-markdown-link-check/commit/0fe4911067fa322422f325b002d2038ba5602170))

mockersf commented:
Could you put this action into our existing CI workflow? https://github.com/bevyengine/bevy/blob/main/.github/workflows/ci.yml
And a check for it by job name in Bors config: https://github.com/bevyengine/bevy/blob/main/.github/bors.toml


For external links
* Could you ignore links to GitHub instead of https links?
* https://crates.io/crates/bevy is a 404 because of https://github.com/rust-lang/crates.io/issues/788. Can you add the header? If not possible, all links to crates.io should be ignored

FlyingRatBull commented:
I added external links to be checked, but for whatever reason, it gets stuck in a redirect loop for `https://developer.android.com/distribute/best-practices/develop/target-sdk`, which I cannot reproduce with curl/firefox:

```
[✖] https://developer.android.com/distribute/best-practices/develop/target-sdk → Status: 0 Error: Exceeded maxRedirects. Probably stuck in a redirect loop https://developer.android.com/distribute/best-practices/develop/target-sdk
    at Redirect.onResponse (/usr/local/lib/node_modules/markdown-link-check/node_modules/request/lib/redirect.js:98:27)
    at Request.onRequestResponse (/usr/local/lib/node_modules/markdown-link-check/node_modules/request/request.js:986:22)
    at ClientRequest.emit (node:events:365:28)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:621:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:127:17)
    at TLSSocket.socketOnData (node:_http_client:487:22)
    at TLSSocket.emit (node:events:365:28)
    at addChunk (node:internal/streams/readable:314:12)
    at readableAddChunk (node:internal/streams/readable:289:9)
    at TLSSocket.Readable.push (node:internal/streams/readable:228:10)
```

Should I explicitly ignore this link in the markdown, in the configuration file or what would be the desired action for this?

I'll move the action into the `ci.yml` soon.

FlyingRatBull commented:
I moved the action into `ci.yml` and referenced it in `bors.toml`.
The total runtime now is ~34 seconds and besides the links in the mentioned PRs and the on commented above, it all runs well.

mockersf commented:
> Should I explicitly ignore this link in the markdown, in the configuration file or what would be the desired action for this?

Not sure why it would fail, it's probably fine to ignore it unless someone else has an idea?

Small nit: I think the config file should go in https://github.com/bevyengine/bevy/tree/main/.github/linters. What do you think?

FlyingRatBull commented:
> Not sure why it would fail, it's probably fine to ignore it unless someone else has an idea?

I ignored it in the Markdown file as opposed to the configuration file because this way, if the link get's removed/changed, the ignore block can be reviewed as well.

> Small nit: I think the config file should go in https://github.com/bevyengine/bevy/tree/main/.github/linters. What do you think?

Good idea. Moved it there and renamed it to `markdown-link-check.json` as well.

FlyingRatBull commented:
Can bors be triggered again? The new check should now succeed, now that the dependent PRs have been merged.

mockersf commented:
you'll need to rebase your PR on main to fix the CI

FlyingRatBull commented:
> you'll need to rebase your PR on main to fix the CI

Thanks, will do

FlyingRatBull commented:
Sorry for the messy git - There was some confusion because of a duplicate broken link.

cart commented:
bors r+
